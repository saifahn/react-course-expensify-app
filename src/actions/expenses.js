import database from '../firebase/firebase'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

/** Under Firebase */
// component calls action generator
// action generator returns function
// component dispatches function
// function runs (has the ability to dispatch other actions an do whatever it wants)

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
})

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData
    const expense = { description, note, amount, createdAt }

    return database
      .ref('expenses')
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          }),
        )
      })
  }
}

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
})

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
})

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
})

export const startSetExpenses = () => {
  return dispatch => {
    return database.ref('expenses').once('value', snapshot => {
      // map the firebase data into an array
      const expenses = []
      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}

// 1. Fetch all expense data once
// 2. Parse that data into an array
// 3. Dispatch SET_EXPENSES
