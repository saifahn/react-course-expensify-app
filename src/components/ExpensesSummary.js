import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectVisibleExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'

// export const ExpensesSummary = props => (
//   <div>
//     {props.expenseCount > 0
//       &&
//       <p>Viewing {props.expenseCount} expense{props.expenseCount > 1
//         && 's'}, totalling {numeral(props.expensesTotal / 100).format('$0,0.00')}
//       </p>}
//   </div>
// );

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')
  return (
    <div>
      <h2>
        Viewing
        {expenseCount} 
{' '}
{expenseWord}, totalling 
{' '}
{formattedExpensesTotal}
      </h2>
    </div>
  )
}

const mapStateToProps = state => {
  const expenses = selectVisibleExpenses(state.expenses, state.filters)
  const expensesTotal = selectExpensesTotal(expenses)
  return {
    expenseCount: expenses.length,
    expensesTotal,
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
