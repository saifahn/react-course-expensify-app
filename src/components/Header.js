import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <div className="header__links">
      <NavLink to="/" activeClassName="is-active" exact>Dashboard</NavLink>
      <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
    </div>
  </header>
);

export default Header;
