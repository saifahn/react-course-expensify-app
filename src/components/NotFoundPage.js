import React from 'react'
import { Link } from 'react-router-dom'

const NotFoundPage = () => (
  <div>
    404 - The page you were looking for {"isn't"} here!{' '}
    <Link to="/">Go home.</Link>
  </div>
)

export default NotFoundPage
