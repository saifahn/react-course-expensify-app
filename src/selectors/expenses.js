import moment from 'moment'

// Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  // filters the expenses array
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt)
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true
      const textMatch =
        typeof text !== 'string' ||
        expense.description.toLowerCase().includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        // the greater value of a/b indicates more recent. 1 will sort b in front of a.
        return a.createdAt < b.createdAt ? 1 : -1
      }
      if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1
      }
    })
}

export default getVisibleExpenses
