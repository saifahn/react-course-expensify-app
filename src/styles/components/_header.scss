.header__links {
  display: flex;
  // justify-content: space-between;
}

.is-active {
  font-weight: bold;
}