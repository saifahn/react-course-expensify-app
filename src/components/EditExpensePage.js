import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense)
    this.props.history.push('/')
  }

  onClick = () => {
    this.props.removeExpense(this.props.expense.id)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <h4>
          Editing the expense with id of
          {this.props.expense.id}
        </h4>
        <button onClick={this.onClick}>Remove</button>
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: id => dispatch(removeExpense({ id })),
})

// props.match.params is the matched URL parameters
// find the expense which has the same ID as the ID of this page
// pass the expense down to EditExpensePage to be accessed in props
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id
    }),
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditExpensePage)
