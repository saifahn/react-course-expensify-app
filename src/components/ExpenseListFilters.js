import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calendarFocused: null,
    };
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    }
    if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          startDateId="startDateId"
          endDateId="endDateId"
          // onDatesChange={({ startDate, endDate }) => this.setState({ startDate, endDate })}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates
          numberOfMonths={1}
          isOutsideRange={() => false}
          // onFocusChange={focusedInput => this.setState({ focusedInput })}
        />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
});

const mapStateToProps = state => (
  {
    filters: state.filters,
  }
);

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
