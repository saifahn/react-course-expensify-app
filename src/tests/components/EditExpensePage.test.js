/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense;
let removeExpense;
let history;
let wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage
    expense={expenses[1]}
    editExpense={editExpense}
    removeExpense={removeExpense}
    history={history}
  />);
});

test('should render EditExpensesPage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  const edited = {
    ...expenses[1],
    description: 'changed',
  };
  wrapper.find('ExpenseForm').prop('onSubmit')(edited);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, edited);
});

test('should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[1].id);
});
