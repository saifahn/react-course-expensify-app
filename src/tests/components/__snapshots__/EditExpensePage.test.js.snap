// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensesPage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Espeon",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <h4>
    Editing the expense with id of
    2
  </h4>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
